dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/MasterSchedulerService.java

```diff
@@ -179,7 +180,8 @@ private void scheduleProcess() throws Exception {
     }
 
     private List<ProcessInstance> command2ProcessInstance(List<Command> commands) {
-        List<ProcessInstance> processInstances = new ArrayList<>(commands.size());
+
+        List<ProcessInstance> processInstances = Collections.synchronizedList(new ArrayList<>(commands.size()));
```

Comment:
The same question.

Reply:
In this method, the thread pool masterPrepareExecService operates on the variable processInstances collection.

![image](https://user-images.githubusercontent.com/37063904/158110881-ccb67815-d8bb-4aa0-8286-52ab2bbb58aa.png)